File: scripts/integrate_assets.js
const fs = require('fs');
const path = require('path');
const { createClient } = require('@supabase/supabase-js');
const AdmZip = require('adm-zip');

// === KONFIGURASI SUPABASE ===
// Baca dari js/supabase-config.js
const configPath = path.join(__dirname, '..', 'js', 'supabase-config.js');
let supabaseUrl, supabaseKey;

try {
    const configContent = fs.readFileSync(configPath, 'utf8');
    const urlMatch = configContent.match(/SUPABASE_URL\s*=\s*["']([^"']+)["']/);
    const keyMatch = configContent.match(/SUPABASE_ANON_KEY\s*=\s*["']([^"']+)["']/);

    if (urlMatch && keyMatch) {
        supabaseUrl = urlMatch[1];
        supabaseKey = keyMatch[1];
        console.log('✅ Ditemukan Supabase URL:', supabaseUrl);
    } else {
        console.error('❌ Tidak ditemukan kredensial di js/supabase-config.js');
        process.exit(1);
    }
} catch (err) {
    console.error('❌ Gagal baca config:', err.message);
    process.exit(1);
}

// Inisialisasi Supabase
const supabase = createClient(supabaseUrl, supabaseKey);
console.log('✅ Supabase client initialized');

// === 1. UPLOAD GAMBAR KE STORAGE ===
async function uploadImages() {
    console.log('\n=== MENGUPLOAD GAMBAR KE STORAGE ===');
    const imagesDir = path.join(__dirname, '..', 'assets', 'langit-korea-images');

    if (!fs.existsSync(imagesDir)) {
        console.log('❌ Direktori', imagesDir, 'tidak ditemukan');
        return;
    }

    const files = fs.readdirSync(imagesDir).filter(f =>
        f.toLowerCase().endsWith('.jpeg') || f.toLowerCase().endsWith('.jpg') || f.toLowerCase().endsWith('.png')
    );
    console.log(`Ditemukan ${files.length} file gambar`);

    for (let i = 0; i < files.length; i++) {
        const filename = files[i];
        const filePath = path.join(imagesDir, filename);

        try {
            const fileBuffer = fs.readFileSync(filePath);
            const contentType = filename.toLowerCase().endsWith('.png') ? 'image/png' : 'image/jpeg';

            const { data, error } = await supabase.storage
                .from('images')
                .upload(filename, fileBuffer, {
                    contentType: contentType,
                    upsert: true
                });

            if (error) {
                console.error(`  ❌ Gagal upload ${filename}:`, error.message);
            } else {
                if (i % 10 === 0) {
                    console.log(`  ✅ Uploaded ${i+1}/${files.length}: ${filename}`);
                }
            }
        } catch (err) {
            console.error(`  ❌ Gagal upload ${filename}:`, err.message);
        }
    }
    console.log(`✅ Selesai upload gambar (${files.length} file)`);
}

// === 2. EXTRACT AUDIO ZIP DAN UPLOAD KE STORAGE ===
async function uploadAudioFromZip() {
    console.log('\n=== MENGEXTRACT AUDIO DARI ZIP DAN UPLOAD ===');
    const zipFiles = [
        path.join(__dirname, '..', 'assets', 'EPS-TOPIK_textbook1_listen.zip'),
        path.join(__dirname, '..', 'assets', 'EPS-TOPIK_textbook2_listen.zip')
    ];

    for (const zipPath of zipFiles) {
        if (!fs.existsSync(zipPath)) {
            console.log(`❌ File ${zipPath} tidak ditemukan, lewati...`);
            continue;
        }

        console.log(`Mengekstrak ${zipPath}...`);
        try {
            const zip = new AdmZip(zipPath);
            const zipEntries = zip.getEntries().filter(entry =>
                entry.name.toLowerCase().endsWith('.mp3')
            );
            console.log(`  Ditemukan ${zipEntries.length} file MP3`);

            for (let i = 0; i < zipEntries.length; i++) {
                const entry = zipEntries[i];
                try {
                    const mp3Buffer = entry.getData();

                    const { data, error } = await supabase.storage
                        .from('audio')
                        .upload(entry.name, mp3Buffer, {
                            contentType: 'audio/mpeg',
                            upsert: true
                        });

                    if (error) {
                        console.error(`  ❌ Gagal upload ${entry.name}:`, error.message);
                    } else {
                        if (i % 5 === 0) {
                            console.log(`  ✅ Uploaded ${i+1}/${zipEntries.length}: ${entry.name}`);
                        }
                    }
                } catch (err) {
                    console.error(`  ❌ Gagal upload ${entry.name}:`, err.message);
                }
            }
            console.log(`✅ Selesai upload audio dari ${zipPath}`);
        } catch (err) {
            console.error(`❌ Gagal ektrak ${zipPath}:`, err.message);
        }
    }
}

// === 3. LOAD DATA JSON DAN INSERT KE TABLE SOAL_EPS ===
async function loadJsonToTable() {
    console.log('\n=== MENGLOAD DATA JSON KE TABLE SOAL_EPS ===');
    const jsonDir = path.join(__dirname, '..', 'assets', 'langit-korea-extracted');

    if (!fs.existsSync(jsonDir)) {
        console.log('❌ Direktori', jsonDir, 'tidak ditemukan');
        return;
    }

    const jsonFiles = fs.readdirSync(jsonDir)
        .filter(f => f.endsWith('.json'))
        .flatMap(f => {
            const filePath = path.join(jsonDir, f);
            try {
                const content = fs.readFileSync(filePath, 'utf8');
                const data = JSON.parse(content);
                return Array.isArray(data) ? data : [];
            } catch (err) {
                console.error(`  ❌ Gagal baca ${f}:`, err.message);
                return [];
            }
        });

    console.log(`Ditemukan ${jsonFiles.length} data soal dari JSON files`);

    let successCount = 0;
    let errorCount = 0;

    for (let i = 0; i < jsonFiles.length; i++) {
        const item = jsonFiles[i];

        // Sesuaikan dengan struktur tabel soal_eps
        const soalData = {
            bab: item.bab || 1,
            tipe: item.tipe || 'membaca',
            teks_soal: item.teks_soal || item.question || '',
            gambar_url: item.gambar_url || null,
            audio_url: item.audio_url || null,
            pilihan_a: item.pilihan_a || item.option_a || '',
            pilihan_b: item.pilihan_b || item.option_b || '',
            pilihan_c: item.pilihan_c || item.option_c || '',
            pilihan_d: item.pilihan_d || item.option_d || '',
            jawaban_benar: item.jawaban_benar || item.correct_answer || '',
            penjelasan: item.penjelasan || '',
            tingkat: item.tingkat || 'mudah',
            akses: item.akses || 'free'
        };

        try {
            const { data, error } = await supabase
                .from('soal_eps')
                .upsert(soalData, { onConflict: 'id' });

            if (error) {
                console.error(`  ❌ Gagal insert soal:`, error.message);
                errorCount++;
            } else {
                successCount++;
                if (i % 10 === 0) {
                    console.log(`  ✅ Inserted ${i+1}/${jsonFiles.length}`);
                }
            }
        } catch (err) {
            console.error(`  ❌ Exception:`, err.message);
            errorCount++;
        }
    }
    console.log(`✅ Selesai load JSON: ${successCount} berhasil, ${errorCount} gagal`);
}

// === FUNGSI UTAMA ===
async function main() {
    console.log('🚀 MULAI INTEGRASI ASET KE SUPABASE');
    console.log('=====================================');

    try {
        // 1. Upload gambar
        await uploadImages();

        // 2. Upload audio dari ZIP
        await uploadAudioFromZip();

        // 3. Load JSON ke table
        await loadJsonToTable();

        console.log('\n✅ INTEGRASI SELESAI!');
        console.log('Cek di Supabase Dashboard:');
        console.log('1. Storage → buckets "images" dan "audio"');
        console.log('2. Table Editor → table "soal_eps"');
    } catch (err) {
        console.error('\n❌ INTEGRASI GAGAL:', err.message);
    }
}

// Jalankan
main();

File: scripts/integrate_assets.py
#!/usr/bin/env python3
"""
Script untuk mengintegrasikan semua aset ke Supabase:
1. Upload gambar ke Storage (bucket: images)
2. Extract audio ZIP dan upload ke Storage (bucket: audio)
3. Load data JSON ke table soal_eps
4. Convert PDF modules ke JSON (opsional)
"""

import os
import json
import zipfile
import io
from supabase import create_client, Client

# === KONFIGURASI SUPABASE ===
# Baca dari js/supabase-config.js
config_path = 'js/supabase-config.js'
supabase_url = None
supabase_key = None

try:
    with open(config_path, 'r') as f:
        content = f.read()
        # Cari URL
        import re
        url_match = re.search(r'SUPABASE_URL\s*=\s*"([^"]+)"', content)
        key_match = re.search(r'SUPABASE_ANON_KEY\s*=\s*"([^"]+)"', content)

        if url_match and key_match:
            supabase_url = url_match.group(1)
            supabase_key = key_match.group(1)
            print(f"✅ Ditemukan Supabase URL: {supabase_url}")
        else:
            print("❌ Tidak ditemukan kredensial di js/supabase-config.js")
            # Gunakan default
            supabase_url = "https://mozmuwrkfsipzfupybwh.supabase.co"
            supabase_key = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Im1veG11d2tmc2lwemZ1cHlid2giLCJyb2xlIjoicm9sZSIsImlhdCI6MTczNjc2MzU1NH0.RPqu-07AyKygnS_bPhMO_IgXSz2r8jkljPc5TGq7Vzg"
except Exception as e:
    print(f"Error baca config: {e}")
    supabase_url = "https://mozmuwrkfsipzfupybwh.supabase.co"
    supabase_key = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Im1veG11d2tmc2lwemZ1cHlid2giLCJyb2xlIjoicm9sZSIsImlhdCI6MTczNjc2MzU1NH0.RPqu-07AyKygnS_bPhMO_IgXSz2r8jkljPc5TGq7Vzg"

print(f"Menggunakan Supabase URL: {supabase_url}")

# Inisialisasi Supabase client
try:
    supabase: Client = create_client(supabase_url, supabase_key)
    print("✅ Supabase client berhasil diinisialisasi")
except Exception as e:
    print(f"❌ Gagal inisialisasi Supabase: {e}")
    exit(1)

# === 1. UPLOAD GAMBAR KE STORAGE ===
print("\n=== MENGUPLOAD GAMBAR KE STORAGE ===")
images_bucket = 'images'
images_dir = 'assets/langit-korea-images'

if os.path.exists(images_dir):
    image_files = [f for f in os.listdir(images_dir) if f.lower().endswith(('.jpeg', '.jpg', '.png'))]
    print(f"Ditemukan {len(image_files)} file gambar")

    for idx, filename in enumerate(image_files):
        filepath = os.path.join(images_dir, filename)
        try:
            with open(filepath, 'rb') as f:
                # Upload ke Supabase Storage
                # Path di bucket: images/filename
                response = supabase.storage.from(images_bucket).upload(
                    filename,  # Path di bucket
                    f.read(),
                    {"content-type": "image/jpeg"}  # Sesuaikan content-type
                )
                if idx % 10 == 0:
                    print(f"  Uploaded {idx+1}/{len(image_files)}: {filename}")
        except Exception as e:
            print(f"  ❌ Gagal upload {filename}: {e}")
    print(f"✅ Selesai upload gambar ({len(image_files)} file)")
else:
    print(f"❌ Direktori {images_dir} tidak ditemukan")

# === 2. EXTRACT AUDIO ZIP DAN UPLOAD KE STORAGE ===
print("\n=== MENGEXTRACT AUDIO DARI ZIP DAN UPLOAD ===")
audio_bucket = 'audio'
zip_files = [
    'assets/EPS-TOPIK_textbook1_listen.zip',
    'assets/EPS-TOPIK_textbook2_listen.zip'
]

for zip_path in zip_files:
    if not os.path.exists(zip_path):
        print(f"❌ File {zip_path} tidak ditemukan, lewati...")
        continue

    print(f"Mengekstrak {zip_path}...")
    try:
        with zipfile.ZipFile(zip_path, 'r') as zip_ref:
            mp3_files = [f for f in zip_ref.namelist() if f.lower().endswith('.mp3')]
            print(f"  Ditemukan {len(mp3_files)} file MP3")

            for idx, mp3_name in enumerate(mp3_files):
                try:
                    # Baca file dari ZIP
                    mp3_data = zip_ref.read(mp3_name)

                    # Upload ke Supabase Storage
                    # Path: audio/mp3_name
                    response = supabase.storage.from(audio_bucket).upload(
                        mp3_name,
                        mp3_data,
                        {"content-type": "audio/mpeg"}
                    )
                    if idx % 5 == 0:
                        print(f"  Uploaded {idx+1}/{len(mp3_files)}: {mp3_name}")
                except Exception as e:
                    print(f"  ❌ Gagal upload {mp3_name}: {e}")
            print(f"✅ Selesai upload audio dari {zip_path}")
    except Exception as e:
        print(f"❌ Gagal ektrak {zip_path}: {e}")

# === 3. LOAD DATA JSON KE TABLE SOAL_EPS ===
print("\n=== MENGLOAD DATA JSON KE TABLE SOAL_EPS ===")
json_dirs = [
    'assets/langit-korea-extracted',
    'assets/langit-korea-json'  # Jika ada
]

for json_dir in json_dirs:
    if not os.path.exists(json_dir):
        print(f"❌ Direktori {json_dir} tidak ditemukan, lewati...")
        continue

    json_files = [f for f in os.listdir(json_dir) if f.endswith('.json')]
    print(f"Ditemukan {len(json_files)} file JSON di {json_dir}")

    for idx, filename in enumerate(json_files):
        filepath = os.path.join(json_dir, filename)
        try:
            with open(filepath, 'r', encoding='utf-8') as f:
                data = json.load(f)

                # Asumsikan data adalah array of objects
                if isinstance(data, list):
                    for item in data:
                        # Sesuaikan dengan struktur tabel soal_eps
                        # Kolom: id, bab, tipe, teks_soal, gambar_url, audio_url, pilihan_a, dst, jawaban_benar, penjelasan, tingkat, akses
                        try:
                            # Cek apakah sudah ada (berdasarkan id atau bab + no)
                            # Insert ke tabel
                            result = supabase.table('soal_eps').insert(item)
                        except Exception as e:
                            print(f"  ❌ Gagal insert {filename}: {e}")
                elif isinstance(data, dict):
                    # Mungkin strukturnya beda
                    print(f"  File {filename} adalah dict, perlu penyesuaian")
        except Exception as e:
            print(f"  ❌ Gagal baca {filename}: {e}")
    print(f"✅ Selesai load JSON dari {json_dir}")

# === 4. KONVERSI PDF KE JSON (OPSIoNAL) ===
print("\n=== KONVERSI PDF MODULES KE JSON ===")
pdf_dir = 'assets/langit-korea-modules'
if os.path.exists(pdf_dir):
    print(f"Direktori {pdf_dir} ditemukan. Untuk mengkonversi PDF ke JSON, jalankan convert_all.py terlebih dahulu.")
else:
    print(f"❌ Direktori {pdf_dir} tidak ditemukan")

print("\n=== INTEGRASI SELESAI ===")
print("Cek di Supabase Dashboard:")
print("1. Storage → buckets 'images' dan 'audio' untuk file-file yang sudah diupload")
print("2. Table Editor → tabel 'soal_eps' untuk data soal")
print("3. Pastikan RLS (Row Level Security) sudah diatur dengan benar")
